core/src/client/mod.rs

```diff
@@ -1299,15 +1304,20 @@ impl KakarotClient for KakarotClientImpl {
         blockhash_opt: Option<H256>,
         blocknum_opt: Option<U256>,
     ) -> Result<BlockTransactions, KakarotClientError> {
-        let mut transactions_vec = vec![];
-        for transaction in initial_transactions {
-            let tx_value = self
-                .starknet_tx_into_kakarot_tx(transaction, blockhash_opt, blocknum_opt)
-                .await;
-            if let Ok(val) = tx_value {
-                transactions_vec.push(val)
-            }
-        }
+        // TODO replace filter_map to handle errors
+        let transactions_vec = stream::iter(initial_transactions)
+            .then(|transaction| async {
+                self.starknet_tx_into_kakarot_tx(transaction, blockhash_opt, blocknum_opt)
+                    .await
+            })
+            .filter_map(|transaction| async {
+                match transaction {
+                    Ok(tx) => Some(tx),
+                    Err(_) => None,
+                }
+            })
+            .collect()
+            .await;
```

Comment:
The core problem with what you are doing is that, although you are using `async` functions (here `stream`), your code is still processed iteratively.

`stream::iter` will yield the items it contains one after the other, exactly like a standard iterator. The main difference is that, if the next item in the stream (which is a future) is not ready yet, the program will wait for it in a non-blocking manner. But it won't go for the item following one until it finished processing the first one.

In your case, there will be no waiting at all because all the items are already available. They all come from a `Vec` stored somewhere in your program memory, so they are immediately available. `stream::iter()` will here behave exactly like a regular `Vec::iter`.

I understand the reason why you want to use it is to be able to use `then`, an `async` function which is not available in the regular `Vec::iter` API. But you actually don't need it

Reply:
What I propose instead is:
1) Create all the `Futures` you intend to execute. (more exactly an iterator over all those `Futures`)
```rust
let handles = initial_transactions
            .into_iter()
            .map(|starknet_transaction| {
                self.starknet_tx_into_kakarot_tx(starknet_transaction, blockhash_opt, blocknum_opt)
            });
```
The trick is to not call `await` on `self.starknet_tx_into_kakarot_tx(starknet_transaction, blockhash_opt, blocknum_opt)`, this way nothing it executed at this point.

2) call [futures::future::join_all](https://docs.rs/futures/latest/futures/future/fn.join_all.html) on your iterator
```rust
futures::future::join_all(handles).await
```
`join_all` will transform all your `Futures` into one single `Future` (the one you await), that will internally wait for all your futures to resolve, but concurrently. Meaning that all the `I/O` requests will be fired more or less "at the same time" and the responses waited at once.

It will yield a `Vec<Result<Transaction, KakarotClientError>>` that we can then normally `iter` and `filter`.

Reply:
I propose you follow the same implementation for the other two functions: `token_balances` and `transaction_receipt` once @Eikix has confirmed that all those futures are safe to be executed concurrently.

Reply:
Ok thanks a lot I understand way better how it works now. But does it mean that the example given [here](https://docs.rs/futures/latest/futures/stream/fn.iter.html) is not really accurate since the vec in the stream is composed of already available items ? I just want to make sure I get it right.